### ISSUE TITLE
[RFC] Include lerna repo as a dependency of another lerna repo

### ISSUE THREAD

**Issue by aleclarson** [lerna/lerna#1834]
I have two forks which are both Lerna repos. One fork depends on the other.

A feature I'm implementing requires changes to both forks. If I want to test these changes, I need a way for the parent fork to import the child fork.

**Does Lerna have a workflow in mind for this already?**

If not, it would be cool if I could tell Lerna to look for packages in the child fork when linking all of the parent fork's packages. 😄

**Comment by evocateur** [comment]
Believe it or not, I definitely get questions fairly frequently (around once a week) about this sort of thing (or something like it) at my day job. It's usually in the context of "I want to develop my React component and its app integration at the same time."

There's nothing wrong with keeping libraries/components alongside applications in the same lerna-managed monorepo, of course. The question of drawing module boundaries is a frequent one both inside and outside of a particular monorepo. Sometimes it's based on ownership, sometimes on problem domain, or in some cases, sheer size of the existing set of modules.

However, I find in most cases that this desire (to link two lerna monorepos directly) is caused combination of two factors: poor unit test patterns & "version shyness" (for lack of a better term). I'll explain:

### Poor Unit Test Patterns

I don't mean to discount the importance of integration/acceptance/end-to-end testing in the least, but I find most solutions to the problem described in this RFC to include better unit test methodology. That is, it _should_ be possible to develop components in one monorepo independently of another monorepo.

The relationship between those two monorepos you describe, if there is one at all, is the package, not the symlink (just like any other npm package). Any changes to one package should be able to unit test their public API to a sufficient granularity to be assured that consumers will be able to use it as expected. Lerna is just glue to make those sorts of changes and improvements easier to a collection of interrelated packages. Lerna is not intended to be a general-purpose replacement for publishing npm packages to a registry.

### Version Shyness

I find very often that folks are extremely reticent to publish new versions of a package, and this reticence can be greatly exacerbated when it's a lerna-managed monorepo with dozens (or hundreds?! 😨) of sibling packages. Maybe it's just an old salt's perspective, but versions are just tarballs, and it's not against the rules to publish several versions (i.e., "fix forward") a day, if necessary. This philosophy works just as well for single modules as it does entire collections of lerna-managed siblings.

Now, I would like to reiterate that solid testing practices (**_not_** "100% test coverage") are absolutely necessary to mitigate risk when it comes to "version diarrhea" (not my term, but it's colorful, neh?). That being said, I tend to believe that it is better to release new versions than to hang back inside endlessly complicated "release" branches and whatnot while waiting for your code to be validated by a consuming application's test process.

If you're using lockfiles (and you _should_), there's no way a newer version of a package gets automatically consumed by any existing users, it's always a manual step. You can even take it a step further and never explicitly publish a new `latest` dist-tag (e.g., `lerna publish --npm-tag next`), and only later (after validation in a test environment, etc) move the dist-tag pointer over to the new version (`npm dist-tag add ${pkg}@${version} latest`). Only then will consumers be informed (by `npm outdated` or greenkeeper-like services) that they might need to upgrade.

## TL;DR:

I'm not currently convinced that this sort of feature is necessary in Lerna.

**Comment by evocateur** [comment]
Thanks for the clarification! I was definitely speaking to a different category of users before.

I think the "simplest" solution in your case is to merge the monorepos. I understand the desire to keep them separate, but it sounds a lot like your development patterns would be better served by a unified approach. (I put scare-quotes around "simplest" because I know from past experience that merging monorepos is anything _but_ simple)

I still believe your request to be outside of the scope of Lerna, mostly because it starts to tread on the territory of package managers like npm and yarn. `yarn link` is even easier to use than its npm cousin (in the context of a monorepo) because it doesn't run an install and break all your lerna-created symlinks.

I can see an argument for making a lerna subcommand that papers over the weirdness of various package managers, but the larger concept of "link two lerna monorepos" is a bridge too far. I would definitely highlight an external tool for this purpose somewhere in Lerna's documentation.

**Comment by aleclarson** [comment]
https://github.com/whitecolor/yalc might be what I'm looking for. I'll let ya know if it solves my use case. 👍

**Comment by evocateur** [comment]
Ooh, that's a neat project. Can't help wondering what "yalc" stands for...

**Comment by aleclarson** [comment]
**Side note:** Merging the two Lerna repos is not an option, since I want to stay in sync with upstream.

**Comment by evocateur** [comment]
Ah, yes, that would defeat my "simple" solution.

**Comment by mrgrain** [comment]
@aleclarson Sorry to dig out this old issue, but have you ever come across a nice solution for this problem? I've run into a similar issue where I'd like the forked dependency to stay in sync with upstream.
